compiler/inference/dataflow.py

```diff
@@ -0,0 +1,107 @@
+from collections.abc import Generator, Iterable, Sequence
+from dataclasses import dataclass
+
+from xdsl.dialects import scf
+from xdsl.ir import Block, Operation, SSAValue, Use
+from xdsl.rewriter import InsertPoint
+from xdsl.traits import IsTerminator
+
+
+@dataclass(frozen=True, slots=True)
+class InsertCandidate:
+    idx: int
+    op: Operation
+
+
+def uses_through_controlflow(val: SSAValue) -> Generator[Use, None, None]:
+    """
+    Find all uses of an SSA value, even through control flow. For example, given the IR:
+
+    ```
+    %a = arith.constant 1 : i32
+
+    %res = scf.for (...) iter_args(%x = %a : i32) {
+        %loop = arith.add ..., %x
+        yield %x
+    }
+
+    arith.add %a, ...
+    ```
+
+    uses_through_controlflow(%a) will yield:
+        - scf.for
+        - %loop = arith.add ..., %x
+        - yield %x
+        - arith.add %a, ...
+    """
+    for use in val.uses:
+        # return each use
+        yield use
+        # if it's an scf.for, recurse on block arg
+        if isinstance(use.operation, scf.For):
+            for_op = use.operation
+            yield from uses_through_controlflow(for_op.body.block.args[use.index - 2])
```

Comment:
I don't understand why we need the -2?
Don't we want something like this?
```suggestion
            yield from uses_through_controlflow(for_op.body.block.args[use.index - 2:])
```

Reply:
Good point, yes. I should add more comments here.

The gist is that we want to get the block arg which represents the input arg, and -2 does that mapping. (`index - 3` is the idx of the loop carried variables, add one back because the first block arg is the loop index, and we end up at `index - 2`)

Reply:
But so now it only works if there is just one block arg, right?

Reply:
I improved/fixed a bunch of stuff, feel free to have a look again!